// angular-pipes/esm2015/array/union.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray, isDeepObject, unwrapDeep, deepIndexOf } from '../utils/utils';
let UnionPipe = class UnionPipe {
    transform(a, b) {
        if ((!isArray(a) && !isDeepObject(a)) || !isArray(b)) {
            return [];
        }
        if (isDeepObject(a)) {
            const unwrapped = unwrapDeep(a);
            if (!isArray(unwrapped)) {
                return [];
            }
            return []
                .concat(unwrapped)
                .concat(b)
                .filter((value, index, input) => deepIndexOf(input, value) === index);
        }
        return []
            .concat(a)
            .concat(b)
            .filter((value, index, input) => input.indexOf(value) === index);
    }
};
UnionPipe = tslib_1.__decorate([
    Pipe({
        name: 'union',
    })
], UnionPipe);
export { UnionPipe };
let NgUnionPipeModule = class NgUnionPipeModule {
};
NgUnionPipeModule = tslib_1.__decorate([
    NgModule({
        declarations: [UnionPipe],
        exports: [UnionPipe],
    })
], NgUnionPipeModule);
export { NgUnionPipeModule };
