// angular-pipes/esm2015/array/uniq.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray, isDeepObject, unwrapDeep, deepIndexOf } from '../utils/utils';
let UniqPipe = class UniqPipe {
    transform(input) {
        if (!isArray(input) && !isDeepObject(input)) {
            return input;
        }
        if (isDeepObject(input)) {
            const unwrappedInput = unwrapDeep(input);
            if (!isArray(unwrappedInput)) {
                return unwrappedInput;
            }
            return unwrappedInput.filter((value, index) => deepIndexOf(unwrappedInput, value) === index);
        }
        return input.filter((value, index) => input.indexOf(value) === index);
    }
};
UniqPipe = tslib_1.__decorate([
    Pipe({
        name: 'uniq',
    })
], UniqPipe);
export { UniqPipe };
let NgUniqPipeModule = class NgUniqPipeModule {
};
NgUniqPipeModule = tslib_1.__decorate([
    NgModule({
        declarations: [UniqPipe],
        exports: [UniqPipe],
    })
], NgUniqPipeModule);
export { NgUniqPipeModule };
