// angular-pipes/esm2015/array/where.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray, isFunction, getProperty } from '../utils/utils';
let WherePipe = class WherePipe {
    /**
     * Support a function or a value or the shorthand ['key', value] like the lodash shorthand.
     */
    transform(input, fn) {
        if (!isArray(input)) {
            return input;
        }
        if (isFunction(fn)) {
            return input.filter(fn);
        }
        else if (isArray(fn)) {
            const [key, value] = fn;
            return input.filter((item) => getProperty(item, key) === value);
        }
        else if (fn) {
            return input.filter((item) => item === fn);
        }
        else {
            return input;
        }
    }
};
WherePipe = tslib_1.__decorate([
    Pipe({
        name: 'where',
    })
], WherePipe);
export { WherePipe };
let NgWherePipeModule = class NgWherePipeModule {
};
NgWherePipeModule = tslib_1.__decorate([
    NgModule({
        declarations: [WherePipe],
        exports: [WherePipe],
    })
], NgWherePipeModule);
export { NgWherePipeModule };
