// angular-pipes/esm2015/array/without.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray, isDeepObject, unwrapDeep, deepIndexOf } from '../utils/utils';
let WithoutPipe = class WithoutPipe {
    transform(input, ...args) {
        if (!isArray(input) && !isDeepObject(input)) {
            return input;
        }
        if (isDeepObject(input)) {
            const unwrappedInput = unwrapDeep(input);
            if (!isArray(unwrappedInput)) {
                return unwrappedInput;
            }
            return unwrappedInput.filter((value) => deepIndexOf(args, value) === -1);
        }
        return input.filter((value) => args.indexOf(value) === -1);
    }
};
WithoutPipe = tslib_1.__decorate([
    Pipe({
        name: 'without',
    })
], WithoutPipe);
export { WithoutPipe };
let NgWithoutPipeModule = class NgWithoutPipeModule {
};
NgWithoutPipeModule = tslib_1.__decorate([
    NgModule({
        declarations: [WithoutPipe],
        exports: [WithoutPipe],
    })
], NgWithoutPipeModule);
export { NgWithoutPipeModule };
