// angular-pipes/esm5/aggregate/group-by.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { getProperty, isArray, isUndefined } from '../utils/utils';
var GroupByPipe = /** @class */ (function () {
    function GroupByPipe() {
    }
    GroupByPipe.prototype.transform = function (input, prop) {
        var e_1, _a;
        if (!isArray(input)) {
            return input;
        }
        var arr = {};
        try {
            for (var input_1 = tslib_1.__values(input), input_1_1 = input_1.next(); !input_1_1.done; input_1_1 = input_1.next()) {
                var value = input_1_1.value;
                var field = getProperty(value, prop);
                if (isUndefined(arr[field])) {
                    arr[field] = [];
                }
                arr[field].push(value);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (input_1_1 && !input_1_1.done && (_a = input_1.return)) _a.call(input_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return Object.keys(arr).map(function (key) { return ({ key: key, value: arr[key] }); });
    };
    GroupByPipe = tslib_1.__decorate([
        Pipe({
            name: 'groupBy',
        })
    ], GroupByPipe);
    return GroupByPipe;
}());
export { GroupByPipe };
var NgGroupByPipeModule = /** @class */ (function () {
    function NgGroupByPipeModule() {
    }
    NgGroupByPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [GroupByPipe],
            exports: [GroupByPipe],
        })
    ], NgGroupByPipeModule);
    return NgGroupByPipeModule;
}());
export { NgGroupByPipeModule };
