// angular-pipes/esm5/aggregate/max.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray } from '../utils/utils';
var MaxPipe = /** @class */ (function () {
    function MaxPipe() {
    }
    MaxPipe.prototype.transform = function (input) {
        if (!isArray(input)) {
            return input;
        }
        if (input.length === 0) {
            return undefined;
        }
        var max = input[0];
        input.forEach(function (value) {
            if (max < value) {
                max = value;
            }
        });
        return max;
    };
    MaxPipe = tslib_1.__decorate([
        Pipe({
            name: 'max',
        })
    ], MaxPipe);
    return MaxPipe;
}());
export { MaxPipe };
var NgMaxPipeModule = /** @class */ (function () {
    function NgMaxPipeModule() {
    }
    NgMaxPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [MaxPipe],
            exports: [MaxPipe],
        })
    ], NgMaxPipeModule);
    return NgMaxPipeModule;
}());
export { NgMaxPipeModule };
