// angular-pipes/esm5/aggregate/min.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray } from '../utils/utils';
var MinPipe = /** @class */ (function () {
    function MinPipe() {
    }
    MinPipe.prototype.transform = function (input) {
        if (!isArray(input)) {
            return input;
        }
        if (input.length === 0) {
            return undefined;
        }
        var min = input[0];
        input.forEach(function (value) {
            if (min > value) {
                min = value;
            }
        });
        return min;
    };
    MinPipe = tslib_1.__decorate([
        Pipe({
            name: 'min',
        })
    ], MinPipe);
    return MinPipe;
}());
export { MinPipe };
var NgMinPipeModule = /** @class */ (function () {
    function NgMinPipeModule() {
    }
    NgMinPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [MinPipe],
            exports: [MinPipe],
        })
    ], NgMinPipeModule);
    return NgMinPipeModule;
}());
export { NgMinPipeModule };
