// angular-pipes/esm5/array/chunk.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray } from '../utils/utils';
var ChunkPipe = /** @class */ (function () {
    function ChunkPipe() {
    }
    ChunkPipe.prototype.transform = function (input, size) {
        if (size === void 0) { size = 1; }
        if (!isArray(input)) {
            return input;
        }
        return [].concat.apply([], input.map(function (_elem, i) {
            return i % size ? [] : [input.slice(i, i + size)];
        }));
    };
    ChunkPipe = tslib_1.__decorate([
        Pipe({
            name: 'chunk',
        })
    ], ChunkPipe);
    return ChunkPipe;
}());
export { ChunkPipe };
var NgChunkPipeModule = /** @class */ (function () {
    function NgChunkPipeModule() {
    }
    NgChunkPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [ChunkPipe],
            exports: [ChunkPipe],
        })
    ], NgChunkPipeModule);
    return NgChunkPipeModule;
}());
export { NgChunkPipeModule };
