// angular-pipes/esm5/array/first-or-default.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray, isFunction, getProperty } from '../utils/utils';
var FirstOrDefaultPipe = /** @class */ (function () {
    function FirstOrDefaultPipe() {
    }
    FirstOrDefaultPipe_1 = FirstOrDefaultPipe;
    FirstOrDefaultPipe.find = function (input, predicate, defaultValue) {
        var found = false;
        var result;
        var i = -1;
        while (!found && ++i < input.length) {
            found = predicate(input[i], i, input);
        }
        if (found) {
            result = input[i];
        }
        if (typeof result === 'undefined' && typeof defaultValue !== 'undefined') {
            result = defaultValue;
        }
        return result;
    };
    FirstOrDefaultPipe.prototype.transform = function (input, predicate, defaultValue) {
        if (!isArray(input)) {
            return input;
        }
        if (isFunction(predicate)) {
            return FirstOrDefaultPipe_1.find(input, predicate, defaultValue);
        }
        else if (isArray(predicate)) {
            var _a = tslib_1.__read(predicate, 2), key_1 = _a[0], value_1 = _a[1];
            return FirstOrDefaultPipe_1.find(input, function (item) { return getProperty(item, key_1) === value_1; }, defaultValue);
        }
        else if (predicate) {
            return FirstOrDefaultPipe_1.find(input, function (item) { return item === predicate; }, defaultValue);
        }
        else {
            return input;
        }
    };
    var FirstOrDefaultPipe_1;
    FirstOrDefaultPipe = FirstOrDefaultPipe_1 = tslib_1.__decorate([
        Pipe({
            name: 'firstOrDefault',
        })
    ], FirstOrDefaultPipe);
    return FirstOrDefaultPipe;
}());
export { FirstOrDefaultPipe };
var NgFirstOrDefaultPipeModule = /** @class */ (function () {
    function NgFirstOrDefaultPipeModule() {
    }
    NgFirstOrDefaultPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [FirstOrDefaultPipe],
            exports: [FirstOrDefaultPipe],
        })
    ], NgFirstOrDefaultPipeModule);
    return NgFirstOrDefaultPipeModule;
}());
export { NgFirstOrDefaultPipeModule };
