// angular-pipes/esm5/array/intersection.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray, isDeepObject, unwrapDeep, deepIndexOf } from '../utils/utils';
var IntersectionPipe = /** @class */ (function () {
    function IntersectionPipe() {
    }
    IntersectionPipe.prototype.transform = function (a, b) {
        if ((!isArray(a) && !isDeepObject(a)) || !isArray(b)) {
            return [];
        }
        if (isDeepObject(a)) {
            var unwrapped = unwrapDeep(a);
            if (!isArray(unwrapped)) {
                return [];
            }
            return unwrapped.reduce(function (intersection, value) {
                return intersection.concat(deepIndexOf(b, value) !== -1 && deepIndexOf(intersection, value) === -1 ? value : []);
            }, []);
        }
        return a.reduce(function (intersection, value) {
            return intersection.concat(b.indexOf(value) !== -1 && intersection.indexOf(value) === -1 ? value : []);
        }, []);
    };
    IntersectionPipe = tslib_1.__decorate([
        Pipe({
            name: 'intersection',
        })
    ], IntersectionPipe);
    return IntersectionPipe;
}());
export { IntersectionPipe };
var NgIntersectionPipeModule = /** @class */ (function () {
    function NgIntersectionPipeModule() {
    }
    NgIntersectionPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [IntersectionPipe],
            exports: [IntersectionPipe],
        })
    ], NgIntersectionPipeModule);
    return NgIntersectionPipeModule;
}());
export { NgIntersectionPipeModule };
