// angular-pipes/esm5/array/range.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
var RangePipe = /** @class */ (function () {
    function RangePipe() {
    }
    RangePipe.prototype.transform = function (_input, size, start, step) {
        if (size === void 0) { size = 0; }
        if (start === void 0) { start = 1; }
        if (step === void 0) { step = 1; }
        var range = [];
        for (var length_1 = 0; length_1 < size; ++length_1) {
            range.push(start);
            start += step;
        }
        return range;
    };
    RangePipe = tslib_1.__decorate([
        Pipe({
            name: 'range',
        })
    ], RangePipe);
    return RangePipe;
}());
export { RangePipe };
var NgRangePipeModule = /** @class */ (function () {
    function NgRangePipeModule() {
    }
    NgRangePipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [RangePipe],
            exports: [RangePipe],
        })
    ], NgRangePipeModule);
    return NgRangePipeModule;
}());
export { NgRangePipeModule };
