// angular-pipes/esm5/array/some.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isArray } from '../utils/utils';
var SomePipe = /** @class */ (function () {
    function SomePipe() {
    }
    SomePipe.prototype.transform = function (input, predicate) {
        if (!isArray(input) || !predicate) {
            return input;
        }
        var result = false;
        var i = -1;
        while (++i < input.length && !result) {
            result = predicate(input[i], i, input);
        }
        return result;
    };
    SomePipe = tslib_1.__decorate([
        Pipe({
            name: 'some',
        })
    ], SomePipe);
    return SomePipe;
}());
export { SomePipe };
var NgSomePipeModule = /** @class */ (function () {
    function NgSomePipeModule() {
    }
    NgSomePipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [SomePipe],
            exports: [SomePipe],
        })
    ], NgSomePipeModule);
    return NgSomePipeModule;
}());
export { NgSomePipeModule };
