// angular-pipes/esm5/math/random.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isNumberFinite } from '../utils/utils';
var RandomPipe = /** @class */ (function () {
    function RandomPipe() {
    }
    RandomPipe.prototype.transform = function (input, min, max) {
        if (min === void 0) { min = 0; }
        if (max === void 0) { max = 1; }
        if (!isNumberFinite(min) || !isNumberFinite(max)) {
            return input;
        }
        if (min > max) {
            max = min;
            min = 0;
        }
        return Math.random() * (max - min) + min;
    };
    RandomPipe = tslib_1.__decorate([
        Pipe({
            name: 'random',
        })
    ], RandomPipe);
    return RandomPipe;
}());
export { RandomPipe };
var NgRandomPipeModule = /** @class */ (function () {
    function NgRandomPipeModule() {
    }
    NgRandomPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [RandomPipe],
            exports: [RandomPipe],
        })
    ], NgRandomPipeModule);
    return NgRandomPipeModule;
}());
export { NgRandomPipeModule };
