// angular-pipes/esm5/string/capitalize.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isString, upperFirst } from '../utils/utils';
var CapitalizePipe = /** @class */ (function () {
    function CapitalizePipe() {
    }
    CapitalizePipe.prototype.transform = function (input, all) {
        if (all === void 0) { all = false; }
        if (!isString(input)) {
            return input;
        }
        if (!all) {
            return upperFirst(input.toLowerCase());
        }
        else {
            return input
                .toLowerCase()
                .split(' ')
                .map(function (value) { return upperFirst(value); })
                .join(' ');
        }
    };
    CapitalizePipe = tslib_1.__decorate([
        Pipe({
            name: 'capitalize',
        })
    ], CapitalizePipe);
    return CapitalizePipe;
}());
export { CapitalizePipe };
var NgCapitalizePipeModule = /** @class */ (function () {
    function NgCapitalizePipeModule() {
    }
    NgCapitalizePipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [CapitalizePipe],
            exports: [CapitalizePipe],
        })
    ], NgCapitalizePipeModule);
    return NgCapitalizePipeModule;
}());
export { NgCapitalizePipeModule };
