// angular-pipes/esm5/string/encode-uri-component.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isString } from '../utils/utils';
var EncodeURIComponentPipe = /** @class */ (function () {
    function EncodeURIComponentPipe() {
    }
    EncodeURIComponentPipe.prototype.transform = function (input) {
        if (!isString(input)) {
            return input;
        }
        return encodeURIComponent(input);
    };
    EncodeURIComponentPipe = tslib_1.__decorate([
        Pipe({
            name: 'encodeURIComponent',
        })
    ], EncodeURIComponentPipe);
    return EncodeURIComponentPipe;
}());
export { EncodeURIComponentPipe };
var NgEncodeURIComponentPipeModule = /** @class */ (function () {
    function NgEncodeURIComponentPipeModule() {
    }
    NgEncodeURIComponentPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [EncodeURIComponentPipe],
            exports: [EncodeURIComponentPipe],
        })
    ], NgEncodeURIComponentPipeModule);
    return NgEncodeURIComponentPipeModule;
}());
export { NgEncodeURIComponentPipeModule };
