// angular-pipes/esm5/string/left-pad.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { leftPad, isString } from '../utils/utils';
var LeftPadPipe = /** @class */ (function () {
    function LeftPadPipe() {
    }
    LeftPadPipe.prototype.transform = function (input, length, character) {
        if (length === void 0) { length = 0; }
        if (character === void 0) { character = ' '; }
        if (!isString(input)) {
            return input;
        }
        return leftPad(input, length, character);
    };
    LeftPadPipe = tslib_1.__decorate([
        Pipe({
            name: 'leftpad',
        })
    ], LeftPadPipe);
    return LeftPadPipe;
}());
export { LeftPadPipe };
var NgLeftPadPipeModule = /** @class */ (function () {
    function NgLeftPadPipeModule() {
    }
    NgLeftPadPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [LeftPadPipe],
            exports: [LeftPadPipe],
        })
    ], NgLeftPadPipeModule);
    return NgLeftPadPipeModule;
}());
export { NgLeftPadPipeModule };
