// angular-pipes/esm5/string/repeat.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isString } from '../utils/utils';
var RepeatPipe = /** @class */ (function () {
    function RepeatPipe() {
    }
    RepeatPipe.prototype.transform = function (input, times, characters) {
        if (times === void 0) { times = 1; }
        if (characters === void 0) { characters = ''; }
        if (!isString(input)) {
            return input;
        }
        if (times <= 0) {
            times = 1;
        }
        var repeated = [input];
        for (var i = 1; i < times; ++i) {
            repeated.push(input);
        }
        return repeated.join(characters);
    };
    RepeatPipe = tslib_1.__decorate([
        Pipe({
            name: 'repeat',
        })
    ], RepeatPipe);
    return RepeatPipe;
}());
export { RepeatPipe };
var NgRepeatPipeModule = /** @class */ (function () {
    function NgRepeatPipeModule() {
    }
    NgRepeatPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [RepeatPipe],
            exports: [RepeatPipe],
        })
    ], NgRepeatPipeModule);
    return NgRepeatPipeModule;
}());
export { NgRepeatPipeModule };
