// angular-pipes/esm5/string/right-pad.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { rightPad, isString } from '../utils/utils';
var RightPadPipe = /** @class */ (function () {
    function RightPadPipe() {
    }
    RightPadPipe.prototype.transform = function (input, length, character) {
        if (length === void 0) { length = 0; }
        if (character === void 0) { character = ' '; }
        if (!isString(input)) {
            return input;
        }
        return rightPad(input, length, character);
    };
    RightPadPipe = tslib_1.__decorate([
        Pipe({
            name: 'rightpad',
        })
    ], RightPadPipe);
    return RightPadPipe;
}());
export { RightPadPipe };
var NgRightPadPipeModule = /** @class */ (function () {
    function NgRightPadPipeModule() {
    }
    NgRightPadPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [RightPadPipe],
            exports: [RightPadPipe],
        })
    ], NgRightPadPipeModule);
    return NgRightPadPipeModule;
}());
export { NgRightPadPipeModule };
