// angular-pipes/esm5/string/strip-tags.pipe.js
// idea from https://github.com/a8m/angular-filter
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isString, isUndefined } from '../utils/utils';
var StripTagsPipe = /** @class */ (function () {
    function StripTagsPipe() {
    }
    StripTagsPipe.prototype.transform = function (input) {
        if (!isString(input) || isUndefined(input))
            return input;
        return input.replace(/<\S[^><]*>/g, '');
    };
    StripTagsPipe = tslib_1.__decorate([
        Pipe({
            name: 'stripTags',
        })
    ], StripTagsPipe);
    return StripTagsPipe;
}());
export { StripTagsPipe };
var NgStripTagsPipeModule = /** @class */ (function () {
    function NgStripTagsPipeModule() {
    }
    NgStripTagsPipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [StripTagsPipe],
            exports: [StripTagsPipe],
        })
    ], NgStripTagsPipeModule);
    return NgStripTagsPipeModule;
}());
export { NgStripTagsPipeModule };
