// angular-pipes/esm5/string/template.pipe.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isString } from '../utils/utils';
var TemplatePipe = /** @class */ (function () {
    function TemplatePipe() {
    }
    TemplatePipe.prototype.transform = function (input) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (!isString(input) || args.length === 0) {
            return input;
        }
        var template = input;
        for (var i = 0; i < args.length; ++i) {
            template = template.replace("$" + (i + 1), args[i]);
        }
        return template;
    };
    TemplatePipe = tslib_1.__decorate([
        Pipe({
            name: 'template',
        })
    ], TemplatePipe);
    return TemplatePipe;
}());
export { TemplatePipe };
var NgTemplatePipeModule = /** @class */ (function () {
    function NgTemplatePipeModule() {
    }
    NgTemplatePipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [TemplatePipe],
            exports: [TemplatePipe],
        })
    ], NgTemplatePipeModule);
    return NgTemplatePipeModule;
}());
export { NgTemplatePipeModule };
