// angular-pipes/esm5/string/truncate.pipe.js
// Inspired from https://github.com/a8m/angular-filter/blob/master/src/_filter/string/truncate.js
import * as tslib_1 from "tslib";
import { Pipe, NgModule } from '@angular/core';
import { isString, isUndefined } from '../utils/utils';
var TruncatePipe = /** @class */ (function () {
    function TruncatePipe() {
    }
    TruncatePipe.prototype.transform = function (input, length, suffix, preserve) {
        if (!isString(input)) {
            return input;
        }
        length = isUndefined(length) ? input.length : length;
        if (input.length <= length) {
            return input;
        }
        preserve = preserve || false;
        suffix = suffix || '';
        var index = length;
        if (preserve) {
            if (input.indexOf(' ', length) === -1) {
                index = input.length;
            }
            else {
                index = input.indexOf(' ', length);
            }
        }
        return input.substring(0, index) + suffix;
    };
    TruncatePipe = tslib_1.__decorate([
        Pipe({
            name: 'truncate',
        })
    ], TruncatePipe);
    return TruncatePipe;
}());
export { TruncatePipe };
var NgTruncatePipeModule = /** @class */ (function () {
    function NgTruncatePipeModule() {
    }
    NgTruncatePipeModule = tslib_1.__decorate([
        NgModule({
            declarations: [TruncatePipe],
            exports: [TruncatePipe],
        })
    ], NgTruncatePipeModule);
    return NgTruncatePipeModule;
}());
export { NgTruncatePipeModule };
